File: problems/680. Valid Palindrome II/680. Valid Palindrome II- sol1.py
class Solution:
    def checkPalindrome(self, s: str, deleted: bool):
        if not s or len(s) == 1:
            return True
        if deleted and s[0] != s[-1]:
            return False
        return self.checkPalindrome(s[1:-1], deleted) if s[0] == s[-1] else (self.checkPalindrome(s[0:-1], True) or self.checkPalindrome(s[1:], True))
    def validPalindrome(self, s: str) -> bool:
        return self.checkPalindrome(s, False)

sol = Solution()
print(sol.validPalindrome('abc'))
print(sol.validPalindrome('aba'))
print(sol.validPalindrome('abca'))
print(sol.validPalindrome('ab'))
print(sol.validPalindrome('acda'))
print(sol.validPalindrome('abdab'))

File: problems/680. Valid Palindrome II/680. Valid Palindrome II- sol1.php
<?php
class Solution {

    /**
     * https://leetcode.com/problems/valid-palindrome-ii/
    * @param String $s
    * @return Boolean
    */
    function validPalindrome($s) {
        list($first, $last) = array(0, strlen($s) - 1);
        while($first < $last) {
            if($s[$first] != $s[$last]) {
                return $this->is_palindrome($s, $first, $last - 1) || $this->is_palindrome($s, $first + 1, $last);
            }
            $first++;
            $last--;
        }
        return true;
    }

    function is_palindrome($s, $i, $j) {
        while($i < $j) {
            if($s[$i] != $s[$j]) {
                return false;
            }
            $i++;
            $j--;
        }
        return true;
    }
}


$sol = new Solution();
echo ($sol->validPalindrome('abc') ? 'true' : 'false') . "\n";
